src/charm.py

```diff
@@ -204,27 +204,42 @@ def _on_install(self, event) -> None:
         except (apt.InvalidSourceError, ValueError, apt.GPGKeyError, URLError):
             self.unit.status = BlockedStatus("couldn't install MongoDB")
 
+        # TODO future PR, this mongod start args is duplicated code and should be put somewhere
+        # else
+
         # Construct the mongod startup commandline args for systemd, note that commandline
         # arguments take priority over any user set config file options. User options will be
         # configured in the config file. MongoDB handles this merge of these two options.
         machine_ip = self._unit_ip(self.unit)
-        mongod_start_args = " ".join(
-            [
-                "ExecStart=/usr/bin/mongod",
-                # bind to localhost and external interfaces
-                "--bind_ip",
-                f"localhost,{machine_ip}",
-                # part of replicaset
-                "--replSet",
-                f"{self.app.name}",
-                "--auth",
-                # keyFile used for authentication replica set peers
-                # TODO: replace with x509
-                "--clusterAuthMode=keyFile",
-                f"--keyFile={KEY_FILE}",
-                "\n",
-            ]
-        )
+        mongod_start_args = [
+            "ExecStart=/usr/bin/mongod",
+            # bind to localhost and external interfaces
+            "--bind_ip",
+            f"localhost,{machine_ip}",
+            # part of replicaset
+            "--replSet",
+            f"{self.app.name}",
+        ]
+
+        # if a new unit is joining a cluster with a legacy relation it should start without auth
+        requires_auth = len(self.client_relations.get_users_from_legacy_relations()) == 0
+        if requires_auth:
```

Comment:
I think you can leave this as `requires_auth =self.client_relations.get_users_from_legacy_relations()`
An empty set will evaluate to False